src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                        if (patternMatch.HasValue)
+                        {
+                            // Since VS item's display text is created as Prefix + DisplayText + Suffix, 
+                            // we can calculate the highlighted span by adding an offset that is the length of the Prefix.
+                            return patternMatch.Value.MatchedSpans.SelectAsArray(s_highlightSpanGetter, matchResult.RoslynCompletionItem);
+                        }
+                    }
+
+                    // If there's no match for Roslyn item's filter text which is identical to its display text,
+                    // then we can safely assume there'd be no matching to VS item's display text.
+                    return ImmutableArray<Span>.Empty;
+                }
+            }
+
+            private FilteredCompletionModel? HandleAllItemsFilteredOut()
+            {
+                if (TriggerReason == CompletionTriggerReason.Insertion)
+                {
+                    // If the user was just typing, and the list went to empty *and* this is a 
+                    // language that wants to dismiss on empty, then just return a null model
+                    // to stop the completion session.
+                    if (_completionRules.DismissIfEmpty)
+                    {
+                        return null;
+                    }
+                }
+
+                // If the user has turned on some filtering states, and we filtered down to
+                // nothing, then we do want the UI to show that to them.  That way the user
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                var selection = UpdateSelectionHint.SoftSelected;
+
+                return new FilteredCompletionModel(
+                    ImmutableArray<CompletionItemWithHighlight>.Empty, selectedItemIndex: 0,
+                    _data.SelectedFilters, selection, centerSelection: true, uniqueItem: null);
+            }
+
+            private ImmutableArray<CompletionFilterWithState> GetUpdatedFilters(IReadOnlyList<MatchResult<VSCompletionItem>> items)
+            {
+                var showCompletionItemFilters = _globalOptions.GetOption(CompletionViewOptions.ShowCompletionItemFilters, _document?.Project.Language);
+                if (!showCompletionItemFilters)
+                    return ImmutableArray<CompletionFilterWithState>.Empty;
+
+                // See which filters might be enabled based on the typed code
+                using var _ = PooledHashSet<CompletionFilter>.GetInstance(out var textFilteredFilters);
+                textFilteredFilters.AddRange(items.SelectMany(n => n.EditorCompletionItem.Filters));
```

Comment:
nit: textFilteredFilters doesn't really mean much to me.  Was there a benefit to this name (as opposed to just 'filters'?).